File: scripts/generate-bbdc-audio.mjs
import { mkdir, stat } from "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";
import bbdcBook from "../src/data/bbdc-yszch.json" with { type: "json" };

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const OUTPUT_DIR = path.resolve(ROOT, process.env.AUDIO_OUTPUT || "generated-audio");
const PYTHON = process.env.PYTHON || (process.platform === "win32" ? "py" : "python3");
const LIMIT = Number.parseInt(process.env.AUDIO_LIMIT || "0", 10);
const ACCENTS = (process.env.ACCENTS || "us,uk")
  .split(",")
  .map((item) => item.trim())
  .filter(Boolean);

const VOICES = {
  us: "en-US-AriaNeural",
  uk: "en-GB-SoniaNeural",
};

function audioFileName(term) {
  return `${encodeURIComponent(String(term).trim().toLowerCase())}.mp3`;
}

function run(command, args) {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, {
      cwd: ROOT,
      stdio: "inherit",
      shell: false,
    });
    child.on("exit", (code) => {
      if (code === 0) resolve();
      else reject(new Error(`${command} ${args.join(" ")} exited with ${code}`));
    });
  });
}

async function exists(file) {
  try {
    await stat(file);
    return true;
  } catch {
    return false;
  }
}

const words = bbdcBook.chapters.flatMap((chapter) => chapter.words);
const uniqueWords = Array.from(new Map(words.map((word) => [word.term.toLowerCase(), word.term])).values());
const selectedWords = LIMIT > 0 ? uniqueWords.slice(0, LIMIT) : uniqueWords;

await mkdir(OUTPUT_DIR, { recursive: true });

for (const accent of ACCENTS) {
  const voice = VOICES[accent];
  if (!voice) {
    console.warn(`Skipping unknown accent: ${accent}`);
    continue;
  }

  const accentDir = path.join(OUTPUT_DIR, accent);
  await mkdir(accentDir, { recursive: true });

  for (const [index, term] of selectedWords.entries()) {
    const output = path.join(accentDir, audioFileName(term));
    if (await exists(output)) continue;

    console.log(`[${accent}] ${index + 1}/${selectedWords.length} ${term}`);
    await run(PYTHON, [
      "-m",
      "edge_tts",
      "--voice",
      voice,
      "--text",
      term,
      "--write-media",
      output,
    ]);
  }
}

File: scripts/generate-bbdc-audio.py
import asyncio
import json
import os
import sys
from pathlib import Path
from urllib.parse import quote

import edge_tts

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

ROOT = Path(__file__).resolve().parents[1]
BOOK_FILE = ROOT / "src" / "data" / "bbdc-yszch.json"
OUTPUT_DIR = Path(os.environ.get("AUDIO_OUTPUT", ROOT / "generated-audio")).resolve()
FAILED_FILE = OUTPUT_DIR / "failed.txt"
LIMIT = int(os.environ.get("AUDIO_LIMIT", "0"))
CONCURRENCY = int(os.environ.get("AUDIO_CONCURRENCY", "6"))
ACCENTS = [item.strip() for item in os.environ.get("ACCENTS", "us,uk").split(",") if item.strip()]

VOICES = {
    "us": "en-US-AriaNeural",
    "uk": "en-GB-SoniaNeural",
}


def audio_file_name(term):
    return f"{quote(term.strip().lower(), safe='')}.mp3"


def load_words():
    with BOOK_FILE.open("r", encoding="utf-8") as file:
        book = json.load(file)
    words = []
    seen = set()
    for chapter in book.get("chapters", []):
        for word in chapter.get("words", []):
            term = str(word.get("term", "")).strip()
            key = term.lower()
            if term and key not in seen:
                seen.add(key)
                words.append(term)
    return words[:LIMIT] if LIMIT > 0 else words


async def generate_one(accent, voice, term, index, total, semaphore):
    output = OUTPUT_DIR / accent / audio_file_name(term)
    if output.exists() and output.stat().st_size > 0:
        return
    output.parent.mkdir(parents=True, exist_ok=True)

    async with semaphore:
        for attempt in range(1, 4):
            try:
                print(f"[{accent}] {index}/{total} {term}")
                communicate = edge_tts.Communicate(term, voice)
                await communicate.save(str(output))
                return
            except Exception:
                if attempt == 3:
                    with FAILED_FILE.open("a", encoding="utf-8") as file:
                        file.write(f"{accent}\t{term}\n")
                    print(f"[{accent}] failed after retries: {term}")
                    return
                await asyncio.sleep(1.5 * attempt)


async def main():
    words = load_words()
    semaphore = asyncio.Semaphore(CONCURRENCY)
    tasks = []
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    if FAILED_FILE.exists():
        FAILED_FILE.unlink()

    for accent in ACCENTS:
        voice = VOICES.get(accent)
        if not voice:
            print(f"Skipping unknown accent: {accent}")
            continue
        for index, term in enumerate(words, 1):
            tasks.append(generate_one(accent, voice, term, index, len(words), semaphore))

    await asyncio.gather(*tasks)


if __name__ == "__main__":
    asyncio.run(main())
